serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/MatrixClient.kt

```diff
@@ -1,45 +1,85 @@
 package xyz.room409.serif.serif_shared
-import kotlinx.coroutines.*
 import io.ktor.client.*
-import io.ktor.client.request.*
-import io.ktor.client.statement.*
 import io.ktor.client.features.json.*
 import io.ktor.client.features.json.serializer.KotlinxSerializer
+import io.ktor.client.request.*
+import io.ktor.client.statement.*
 import io.ktor.http.*
+import kotlinx.coroutines.*
+import kotlin.concurrent.thread
+import kotlin.synchronized
 
-sealed class Outcome<out T : Any> { }
+sealed class Outcome<out T : Any>
 data class Success<out T : Any>(val value: T) : Outcome<T>()
 data class Error(val message: String, val cause: Exception? = null) : Outcome<Nothing>()
 
 // Returns the first Event of class T, or null
 inline fun <reified T> List<Event>.firstOfType(): T? = this.map { (it as? T?) }.firstOrNull { it != null }
+// Returns the content of type T from the first Event of type StateEvent<T>
+// This can't just be a call to firstofType<StateEvent<T>> because the safe cast operator only works one level, and you'd
+// still get ClassCastExceptions when it returns an Event that is indeed of type StateEvent<?> but not instantiated with
+// the right T. Java generics are bad, yall, and Kotlin in trying to be nice sometimes makes things much worse.
+inline fun <reified T> List<Event>.firstStateEventContentOfType(): T? = this.map { ((it as? StateEvent<T>?)?.content) as? T? }.firstOrNull { it != null }
 
-class MatrixSession(val client: HttpClient, val access_token: String, var transactionId : Long) {
-    var sync_response: SyncResponse? = null
+class MatrixSession(val client: HttpClient, val access_token: String, var transactionId: Long, val onSync: () -> Unit) {
+    private var sync_response: SyncResponse? = null
+    private var sync_should_run = true
+    private var sync_thread: Thread? = thread(start = true) {
+        val timeout_ms: Long = 10000
+        var fail_times = 0
+        while (sync_should_run) {
```

Comment:
Should we add a sleep on this loop regardless of an exception occurring instead of spamming the server with requests? And not an issue/blocker for this PR but I assume this will need to be handled differently for polling updates on android (and presumably iOS), or at least we will likely need to provide some sort of notify callback?

Reply:
It won't spam! The server is designed for long-polling - it intentionally doesn't respond until there's more data, at which point it returns that new data. You can then make another request which will wait until there's more data, etc

Reply:
Also, there is a notify callback - onSync!